src/web/TextInput.tsx

```diff
@@ -43,6 +43,88 @@ export interface TextInputContext {
     focusArbitrator?: FocusArbitratorProvider;
 }
 
+interface TextInputPlaceholderCacheItem {
+    refCounter: number;
+    styleElement: HTMLStyleElement;
+}
+
+class TextInputPlaceholderSupport {
+    private static _cachedStyles: { [color: string]: TextInputPlaceholderCacheItem } = {};
+
+    static getClassName(color: string): string {
+        const key = this._colorKey(color);
+
+        return `reactxp-placeholder-${key}`;
+    }
+
+    static addRef(color: string) {
+        if (typeof document === undefined) {
+            return;
+        }
+
+        const cache = this._cachedStyles;
+        const key = this._colorKey(color);
+
+        if (cache.hasOwnProperty(key)) {
+            cache[key].refCounter++;
+        } else {
+            const className = this.getClassName(color);
+            const style = document.createElement('style');
+            style.type = 'text/css';
+            style.textContent = this._getStyle(className, color);
+
+            document.head.appendChild(style);
+
+            cache[key] = {
+                refCounter: 1,
+                styleElement: style,
+            };
+        }
+    }
+
+    static removeRef(color: string) {
+        const cache = this._cachedStyles;
+        const key = this._colorKey(color);
+
+        if (cache.hasOwnProperty(key)) {
+            const item = cache[key];
+
+            if (--item.refCounter < 1) {
+                const styleElement = item.styleElement;
+                if (styleElement.parentNode) {
+                    styleElement.parentNode.removeChild(styleElement);
+                }
+                delete cache[key];
+            }
+        }
+    }
+
+    private static _colorKey(color: string): string {
+        return color.toLowerCase()
```

Comment:
I am not particularly happy about this part and it's really hard to support all of the edge cases. If you have a better idea how to handle, I am all ears. One of solutions I had in mind was to use a proxy HTMLElement to filter colors through `style.color`, at least WebKit seems to return normalised colors back. Not sure how this behaviour varies across platforms though.

```
var e = document.createElement('span')
e.style.color = '#ccc'
e.style.color // -> rgb(204, 204, 204)
// the downside is, predefined colors like red do not get normalized to rgb(..) but remain the same
```

The other solution would be to rely on some CSS color normalisation library.

Reply:
I don't think it's necessary to normalize colors. If an app uses color names, they tend to use them everywhere and vice versa. In the case where they intermix names and numeric colors, it will just be slightly less efficient.